src/cargo/core/resolver/mod.rs

```diff
@@ -256,12 +257,43 @@ impl<'a> Iterator for DepsNotReplaced<'a> {
     }
 }
 
+enum RcList<T> {
+    Data(Rc<(T, RcList<T>)>),
+    Empty,
+}
+
+impl<T> RcList<T> {
+    fn new() -> RcList<T> {
+        RcList::Empty
+    }
+
+    fn push(&mut self, data: T) {
+        let node = Rc::new((data, mem::replace(self, RcList::Empty)));
+        *self = RcList::Data(node);
+    }
+}
+
+// Not derived to avoid `T: Clone`
+impl<T> Clone for RcList<T> {
+    fn clone(&self) -> RcList<T> {
+        match *self {
+            RcList::Data(ref d) => RcList::Data(d.clone()),
+            RcList::Empty => RcList::Empty,
+        }
+    }
+}
+
+enum GraphNode {
+    Add(PackageId),
+    Link(PackageId, PackageId),
+}
+
 #[derive(Clone)]
 struct Context<'a> {
-    activations: HashMap<(String, SourceId), Vec<Rc<Summary>>>,
-    resolve_graph: Graph<PackageId>,
+    activations: HashMap<String, HashMap<SourceId, Vec<Summary>>>,
```

Comment:
:+1: (at least, the branch is named hamt :) )

Just to clarify, I wanted to use `SmallMap` just for the inner map here. Hm, and I now actually think that the innermost `Vec` is what should be `Small`. If it is correct that the number of activations is almost always one, we can try to reduce the number of allocations here by storing the first activation on the stack, and this is somewhat orthogonal to making the whole thing a persistent data structure.